### PR TITLE
ClipQKV

### PR DESCRIPTION
**Pull request by vchiley**
This PR
- enables `qk_ln` in triton attn variant
- enables clipping qkv before attn is applied

On its own, Alibi does not fully solve the stability issue; when clipping is added to qkv, training becomes more stable.
Screenshot 2023-02-27 at 9 10 46 PM

Without Alibi, clipping qkv is still very stable (it just doesn't do as well)
Screenshot 2023-02-27 at 9 12 12 PM

I tried clipping values of `{0.1, 1, 2, 10}`; `{0.1, 1}` were too aggressive; `{2, 10}` are shown. We can see that when the clipping value is higher (`10`) the network is slightly less stable and gets psuedo-loss spikes from which it recovers gracefully.

ClipQKV+Alibi outperforms QKLN
Screenshot 2023-02-27 at 9 18 20 PM

ClipQKV+Alibi has the exact same performance as QKLN+Alibi
Screenshot 2023-02-27 at 9 19 07 PM

### CODE REVIEW

File: examples/llm/src/models/layers/attention.py
Comment (abhi-mosaic) [inline examples/llm/src/models/layers/attention.py:272]: if using `clamp_` do you need to do the assignment?

File: examples/llm/src/models/mosaic_gpt.py
Comment (abhi-mosaic) [inline examples/llm/src/models/mosaic_gpt.py:40]:
```suggestion
        if cfg.get('attn_qk_ln') and cfg.attn_impl not in ['flash', 'triton']:
```

**Review by abhi-mosaic — COMMENTED** [review]
Gonna leave this to others to review + approve, but could I suggest adding some unit tests to check equality between the torch/causal/triton attention blocks?

File: examples/llm/src/models/layers/attention.py
Comment (vchiley) [reply examples/llm/src/models/layers/attention.py]: making valid points!

**Comment by vchiley** [comment]
> adding some unit tests to check equality between the torch/causal/triton attention blocks?
https://mosaicml.atlassian.net/browse/RESEARCH-468

**Review by dakinggg — APPROVED** [review]
LGTM

**Comment by dakinggg** [comment]
Looks like the tests got OOM killed. Try turning down some of the params (e.g. model size, sequence length) on the test that got killed (https://github.com/mosaicml/examples/blob/d1a855f4ac0bc4bc6ec218b6748f35fa008eafbf/examples/llm/tests/test_model.py#L179)

**Comment by vchiley** [comment]
https://github.com/mosaicml/examples/pull/197#issuecomment-1449479110

yeah they're gpu tests. I need to xfail them on cpu. Forgot to do that...

File: examples/llm/src/models/layers/attention.py
Comment (vchiley) [reply examples/llm/src/models/layers/attention.py:238]: not causal unless you pass `causal=True` [here](https://github.com/mosaicml/examples/pull/197/files/212b6926e355b3f6be20acbdf1c80f735d448435#diff-22824d61d86c454e9e92d5ea2b1a233d8d972160e01d2261f3dd8f18c307c4f5R257).
At this point its still generic attn.

This is also true for the cuda version of flash attn, probably should have noted it [here](https://github.com/mosaicml/examples/pull/174)

@bcui19 should we make both variants be `causal_attn` or `inner_attn`?

I was following convention [here](https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attention.py#L90).

File: examples/llm/src/models/layers/attention.py
Comment (vchiley) [reply examples/llm/src/models/layers/attention.py:239]: actually just adding `cfg.get('softmax_scale')` enables users to set sm scale (in a manual way but still enables it).
I'll add that; it'll default to `None`.

File: examples/llm/src/models/layers/attention.py
Comment (bcui19) [reply examples/llm/src/models/layers/attention.py:238]: Yeah it sounds like it makes sense to name both of them `inner_attn` just match conventions.

**Review by bcui19 — APPROVED** [review]
LGTM after gating the tests to be on GPU only

**Review by codestar12 — APPROVED** [review]
Looks good!

**Pull request merged**
